bson/fuzz_test.go

```diff
@@ -0,0 +1,34 @@
+package bson
+
+import (
+	"testing"
+)
+
+func FuzzDecode(f *testing.F) {
+	seedBSONCorpus(f)
+
+	f.Fuzz(func(t *testing.T, data []byte) {
+		for _, typ := range []func() interface{}{
+			func() interface{} { return new(D) },
+			func() interface{} { return new([]E) },
+			func() interface{} { return new(M) },
+			func() interface{} { return new(interface{}) },
+			func() interface{} { return make(map[string]interface{}) },
+			func() interface{} { return new([]interface{}) },
+		} {
+			i := typ()
+			if err := Unmarshal(data, i); err != nil {
+				return
+			}
+
+			encoded, err := Marshal(i)
+			if err != nil {
+				t.Fatal("failed to marshal", err)
+			}
+
+			if err := Unmarshal(encoded, i); err != nil {
```

Comment:
Ah interesting that we unmarshal, marshal and unmarshal _again_. What's the reasoning there?

Reply:
The first unmarshal is to check the validity of the extended JSON. We only want this to generate corpus data on a panic/crash. If the value is not valid BSON and we don't crash, then this subtest if over.

The first marshal is to check that we can encode valid data structures into BSON, it seems unnecessary to fuzz the encoding of invalid data structures.

The last unmarshal is where we check that we can decode valid BSON, if this fails/panics/crashes then we want to know about it.